File: 785-is-graph-bipartite/785-is-graph-bipartite.cpp
class Solution {
    
    void printList(vector<int> v){
        for(auto a: v){
            cout << a << "  ";
        }
        cout << endl;
    }
    
    
public:
    bool isBipartite(vector<vector<int>>& graph) {
        
        queue<int> q;
        vector<int> colors(graph.size(), -1);
        vector<int> visited(graph.size(), -1);
        
        for(int i = 0; i < graph.size(); i++){
            if(colors[i] == -1){
                q.push(i);
                colors[i] = 1;
                int parentColor = 1;
                while(!q.empty()){
                    int top = q.front();
                    q.pop();
                    for(auto node: graph[top]){
                        if(colors[node]==-1){
                            // cout << node << " ." << top << endl;
                            // printList(colors);
                            q.push(node);
                            colors[node] = 1 - colors[top];
                        }else{
                            // cout << node << " " << top << endl;
                            // printList(colors);
                            if(colors[node] == colors[top]) return false;
                            
                        }
                    }
                }
            }
        }
        
        return true;
    }
};

File: 785-is-graph-bipartite/785-is-graph-bipartite.py
from collections import deque
class Solution:
    
    def solveBfs(self, graph):
        Q = deque()
        n = len(graph)
        visited = [-1 for i in range(n+1)]
        
        
        for i in range(n):
            if visited[i] == -1:
                Q.append(i)
                visited[i] = 0

                while len(Q):
                    node = Q.popleft()

                    for nbr in graph[node]:
                        if visited[nbr] == -1:
                            visited[nbr] = 1 - visited[node]
                            Q.append(nbr)
                        elif visited[nbr] == visited[node]:
                            return False
                
        return True
        
        
        
    
    def isBipartite(self, graph: List[List[int]]) -> bool:
        return self.solveBfs(graph)
        
        